### DISCUSSION TITLE
Change mqtt topic in node

### DISCUSSION THREAD

**Discussion by hakha4** [timmbogner/Farm-Data-Relay-System#191] [General]
Hi!
I have a Node sending relaystatus as a float value. My problem is that I collect data in Home Assistant and it seems more or less impossible to create a sensor for every inudividual bit in value.  In Node sketch I have split the value to bit true/false. So my question is, can topic be changed from dfrs/data to  dfrs/data/relay1, dfrs/data/relay2 etc. So I can send a specific topic for every relay.

 I know `loadFDRS(data,STATUS_T); ` don't include the option to select topic,but is there a workaround ?

Best Regards Håkan

**Comment by timmbogner** [comment]
Hello Håkan!
Could you post the code for the node that you're working with? I think then I'll have a better idea of what's up.

There's another way to use loadFDRS() where you tell it what ID you want to use, could that be of help? (`void loadFDRS(float d, uint8_t t, uint16_t id`)

I am still planning to fix that float issue soon!

**Comment by hakha4** [comment]
Hi!

Code:

``` cpp
/*
MCP23017 uses inverted logic. Output from 0 to 15
To set relay ON or OFF convert short integer to 'binary'  and set each output according
To set  relay 1,3 and 4 to ON (=0) and the rest to OFF (=1)
convert 'binary representation 1111111111110100 = 65524
SEND:
[{"id":31,"type":0,"data":65524}]

compile with DOIT ESP32 DEVKIT V1



*/
#define FDRS_DEBUG
#define READING_ID 31
#define GTWY_MAC 0x01
#define USE_ESPNOW
#define status_PIN 2

#include <fdrs_node.h>
#include <Adafruit_MCP23X17.h>


 
Adafruit_MCP23X17 mcp;

// for debug only
String str_binary = "";
/*### incoming data variable ###
make sure all relays are set to OFF if node restarts for safety
1111111111111111 = 65535
*/
int tmpdata = 65535;

//########## Non blocking delay var's #################
unsigned long previousMillis = 0;
const long interval = 2000;

//only send if value have  changed not to stress system, causes TIMEOUT !!!!
//int previousValue = 0;
//int currentValue = 0;

//##############################
void fdrs_recv_cb(DataReading theData) {
  digitalWrite(status_PIN,LOW);//for visual debug in field
  tmpdata = (unsigned short)theData.d;
  str_binary = "";
  String tmpbit;
//#### write new values to MCP23017 #####
//     read bitvalue at each position and write value to MCP23017 
  for(int i = 15; i >= 0; i--) {
    // check bits, LSB first
    int bitValue = bitRead(tmpdata, i); // read the bit at position i
    tmpbit = String(bitValue);//save result for debug
     mcp.digitalWrite(i, bitValue);//write value to MCP23017
     str_binary = str_binary + tmpbit;//add in debugstring
	 
//******************  If possible set topic to fdrs/data/relay(i) with value bitvalue ************

  }
  
}
void setup() {
 pinMode(status_PIN, OUTPUT);//visualizing node ok
 
  beginFDRS();
  if (addFDRS(1000, fdrs_recv_cb))
    subscribeFDRS(READING_ID);
Wire.begin(); // I/O expander

  mcp.begin_I2C();
//set all pins to output
for (int i = 0; i <= 15; i++) {
mcp.pinMode(i, OUTPUT);

}

}
void loop() {

 

 unsigned long currentMillis = millis();
//currentValue = tmpdata;

 loopFDRS();
 //avoid stressing system, only send when new value arrives
//if (currentValue != previousValue) {


if (currentMillis - previousMillis >= interval) {
    // save the last time when code was running
    previousMillis = currentMillis;
	
//******************  If possible set topic to fdrs/data/relayx with payload bitvalue (0/1)************
   //send data
   loadFDRS(tmpdata,STATUS_T); 
  
  //  DBG(sendFDRS());
    if(sendFDRS()){
           //LED blinking: alive,sending OK
             digitalWrite(status_PIN,HIGH);
      Serial.println(str_binary);       
      DBG("TMPDATA ");
      DBG(tmpdata);
    
    } else {
      //LED NOT blinking: ERROR of some kind
      digitalWrite(status_PIN,LOW);
      DBG("ERROR SENDING DATA");
    }

}//end if currentmillis...
//}//end if current value
 //previousValue = currentValue;
//sleepFDRS(10);  //Sleep time in seconds
}//end void loop
```


================================================================`
For now I just send a testvalue (tmpdata) and it works perfect receiving and sending data via mqtt.
But, since it's so complicated to split a float to individual sensor's in HA one way is to send every bit at a time with a _subtopic_ so it can be identified in HA. That is instead of one topic for this node it vill have 16

I'm not sure what "There's another way to use loadFDRS() where you tell it what ID you want to use, could that be of help? (void loadFDRS(float d, uint8_t t, uint16_t id)" will do ? Still the float value to be sent,right?


Best Regards Håkan
PS can't get code formatting to work (<>), sorry

**Reply by timmbogner** [reply]
Okay, I think I see now. 

In your current usage you're relying on the READING_ID to assign an individual ID to your board and then basing everything off of that. You actually can assign a separate ID to each relay, and use a GET/SET paradigm to control them.

I think my [Irrigation example](examples/Controller_examples/Irrigation/Irrigation.ino) is what you need, although you will need to expand it from 4 to 8 relays. You'll also have to make it feed binary to your IO expander of course. To operate the sketch, you will use the DataReading 'type' to denote whether you'd like to set (0) or get (1) a value. If you are setting a value, set the DataReading's 'data' payload to a 1 or 0. It also sends a readout of the states of all of the relay pins every time it changes one (easily disabled). Is this setup copacetic with Home Assistant?

Soon I hope to support full binary, so this sketch could be given the option of sending the relay states all under one ID. I think both ways may be useful to different folks.

**Reply by hakha4** [reply]
Hi!
Thank's for answer.
 The link to irrigation example is broken and I cant find it anywhere , and I'm not 100% sure how to deal with multiple ID's, in one Node. (doesn't the gateway get confused ??) Can You point me where to find the example so I can study this in detail?
Thank's Håkan

**Reply by timmbogner** [reply]
Sorry about that. Here's the link, it's in examples/Controller_examples/

https://github.com/timmbogner/Farm-Data-Relay-System/tree/main/examples/Controller_examples/Irrigation

**Reply by timmbogner** [reply]
> I'm not 100% sure how to deal with multiple ID's, in one Node. (doesn't the gateway get confused ??)


It's really as simple as just adding an extra argument to loadFDRS() defining what you'd like your ID to be.
All the gateway cares about is the source of 
the data. Any data that comes from an unknown ESP-NOW address (MAC) is assumed to be incoming (sensor) data and is forwarded using the instructions defined by ESPNOWG_ACT in the gateway config. Gateways never actually look at the IDs of readings.

**Reply by hakha4** [reply]
Hi!

> It's really as simple as just adding an extra argument to loadFDRS() defining what you'd like your ID to be.

Nothing is simple for me it seems!

Anyway,I made a 'skeleton' of Your code just to see If subscription is ok BUT with this code

``` cpp
#define FDRS_DEBUG
#define READING_ID 32//Compile error  due to DBG macro if commented out !!
#define GTWY_MAC 0x01
#define USE_ESPNOW
#include <fdrs_node.h>

#define CONTROL_1 101  //Address for controller 1
#define CONTROL_2 102  //Address for controller 2
#define CONTROL_3 103  //Address for controller 3
#define CONTROL_4 104  //Address for controller 4



bool isData = false;
bool newStatus = false;

void fdrs_recv_cb(DataReading theData) {
  DBG(String(theData.id));
  DBG(String(theData.t));
}

void setup() {
   beginFDRS();
  pingFDRS(1000);
  if (addFDRS(1000, fdrs_recv_cb)) {
    subscribeFDRS(CONTROL_1);
    subscribeFDRS(CONTROL_2);
    subscribeFDRS(CONTROL_3);
    subscribeFDRS(CONTROL_4);
} else {
    DBG("Not Connected");
}

}

void loop() {
 loopFDRS();
  if (isData) {
    isData = false;
    DBG("Update relays here")
  }
  if (newStatus) {
    newStatus = false;
    if (sendFDRS()) {
      DBG("Packet received by gateway");
    } else {
      DBG("Unable to communicate with gateway!");
    }
  }
}
```
I get this output in Serial monitor 

**Gateway: 1
Initializing ESP-NOW!
 ESP-NOW Initialized.
 ESP-NOW ping sent.
No ESP-NOW ping returned within 1000ms.
ESP-NOW peer registration request submitted to 1
No gateways accepted the request
Not Connected
Refreshing registration to 1
No gateways accepted the request**

If I comment out `#define READING_ID 32//Compile error  due to DBG macro if commented out !!`
It Won't compile due to error in DBG macro

_In file included from c:\Users\hakan\OneDrive\Documents\Arduino\libraries\Farm-Data-Relay-System-main\src/fdrs_node.h:58,
                 from C:\Users\hakan\AppData\Local\Temp\.arduinoIDE-unsaved2024026-1700-lu8b4j.k7uo\sketch_jan26b\sketch_jan26b.ino:6:
c:\Users\hakan\OneDrive\Documents\Arduino\libraries\Farm-Data-Relay-System-main\src/fdrs_node.h: In function 'void beginFDRS()':
c:\Users\hakan\OneDrive\Documents\Arduino\libraries\Farm-Data-Relay-System-main\src/fdrs_node.h:86:31: error: 'READING_ID' was not declared in this scope
   DBG(" Reading ID " + String(READING_ID));
                               ^~~~~~~~~~
c:\Users\hakan\OneDrive\Documents\Arduino\libraries\Farm-Data-Relay-System-main\src/fdrs_debug.h:6:31: note: in definition of macro 'DBG'
 #define DBG(a) Serial.println(a);
                               ^
c:\Users\hakan\OneDrive\Documents\Arduino\libraries\Farm-Data-Relay-System-main\src/fdrs_node.h:86:31: note: suggested alternative: 'PENDING'
   DBG(" Reading ID " + String(READING_ID));
                               ^~~~~~~~~~
c:\Users\hakan\OneDrive\Documents\Arduino\libraries\Farm-Data-Relay-System-main\src/fdrs_debug.h:6:31: note: in definition of macro 'DBG'
 #define DBG(a) Serial.println(a);
                               ^
c:\Users\hakan\OneDrive\Documents\Arduino\libraries\Farm-Data-Relay-System-main\src/fdrs_node.h: In function 'void loadFDRS(float, uint8_t)':
c:\Users\hakan\OneDrive\Documents\Arduino\libraries\Farm-Data-Relay-System-main\src/fdrs_node.h:229:23: error: 'READING_ID' was not declared in this scope
   DBG("Id: " + String(READING_ID) + " - Type: " + String(t) + " - Data loaded: " + String(d));
                       ^~~~~~~~~~
c:\Users\hakan\OneDrive\Documents\Arduino\libraries\Farm-Data-Relay-System-main\src/fdrs_debug.h:6:31: note: in definition of macro 'DBG'
 #define DBG(a) Serial.println(a);
                               ^
c:\Users\hakan\OneDrive\Documents\Arduino\libraries\Farm-Data-Relay-System-main\src/fdrs_node.h:229:23: note: suggested alternative: 'PENDING'
   DBG("Id: " + String(READING_ID) + " - Type: " + String(t) + " - Data loaded: " + String(d));
                       ^~~~~~~~~~
c:\Users\hakan\OneDrive\Documents\Arduino\libraries\Farm-Data-Relay-System-main\src/fdrs_debug.h:6:31: note: in definition of macro 'DBG'
 #define DBG(a) Serial.println(a);
                               ^

exit status 1

Compilation error: exit status 1_


Are the irrigation example a Lora node?
With espnow it doesn't seems to work.

I'm in a moment 22 just now

_**It's really as simple as just adding an extra argument to loadFDRS() defining what you'd like your ID to be.**_
Do You by this mean that I could use my working node and add an extra argument like this?

`loadFDRS(valueToSend,MYNUMBER, STATUS_T);`

Sorry to give you so much work with this but I'm really impressed over how well it works on my for now 2 Lora nodes and 1 espnow node and really wan't to expand with my own ideas

Best Regards Håkan

**Comment by hakha4** [comment]
Hi again!
Disregard my last post from examples and checking code I see that READING_ID must be present even if one define own  subscribeFDRS
I've think I figured out what you mean by adding extra argument to loadFDRS().

One  'dirty' way is to send 'Type' as o to 15 indicating what pin numer is actual and as data either 0 or 1. This I can catch in Home assistant.

I have two lora nodes and my relay node running and data comes in just fine in Home assistant via mqtt. I have tested to send from mqtt explorer and data changes in Home assistant bur I realize that data is not recieved in relay node! Is it correct to senda data TO node in json like [{"id":31,"type":0,"data":65524}] or is it another format when publishing to node ?? I tried to see from code but I can't really find out.

So in summary have I understood you right with extra argument ?
Can you point me how to send message TO node correct ?
Sorry to ask you so many questions !
Best Regards Håkan

**Reply by timmbogner** [reply]
Ooooh I see. At the moment you can't comment out READING_ID, but you can set it to zero and/or ignore it. Soon I will change it to make it so that you can comment it out.
You're right, that's a quick and easy way to do it. Especially in controllers you are free to use the ID, type, and data elements as you wish. The Irrigation sketch just has some useful features already built-in.

Referencing[ this section](https://github.com/timmbogner/Farm-Data-Relay-System/blob/main/extras/Node.md#sensor-api) of the documentation:

>**loadFDRS(float data, uint8_t type, uint16_t id);**
Loads some data into the current packet. 'data' is a float, 'type' is the data type (see below), and 'id' is the DataReading id.

Here is a snippet of Irrigation.ino to illustrate. This is the portion where the controller is reading the state of the relay, then building a packet to send back to the gateway:
``` cpp
        case CONTROL_1:    // CONTROL_1 is the ID (address) of your first relay (101 by default)
          if (digitalRead(COIL_1) == HIGH) {   // check COIL_1, which is the pin number of your first relay
            loadFDRS(1, STATUS_T, CONTROL_1);   // If it's high, load a packet with data "1", type "0", and address "101"
          } else {
            loadFDRS(0, STATUS_T, CONTROL_1);   // STATUS_T is a constant and is defined as "0"
          }
          break;
 ```
 I think you're on the right track with sending to a node... just enter it as JSON in the serial monitor of the UART gateway. Your routing options, specifically `SERIAL_ACT` will tell the gateway where to send the data next.
 
Lastly, to add a small line of code ```like(this)``` in a comment you surround the text with single ` symbols

To comment an entire block of code you add three of those symbols in a single line to start the block and three more to end it. You can type "cpp" after the first set to format it for C++. Click edit on you previous comment and you'll see how I did it. For a block of text, the tags need to be on their own line.

Thanks for your patience!

**Reply by hakha4** [reply]
Thank's for YOUR patience !!. I'm ashamed taking so much of your time. 🥇 

` DBG("Recieved");` to debug if message recieved but newer print's out.


```cpp
//##############################
void fdrs_recv_cb(DataReading theData) {
  DBG("Recieved");
// check if mqtt message is ever receided for debug
  tmp_pinnumber = theData.t;
  tmpdata = (unsigned short)theData.d;
 
  
  }
```
I'm a bit confused about routing. In my gateway I used the original settings from example, is something missing for sending mqtt to nodes ?

```cpp

#define FDRS_DEBUG
#define READING_ID 4
#define GTWY_MAC 0x01
#define USE_ESPNOW
#define status_PIN 2

// Routing
// Options: sendESPNowNbr(1 or 2); sendESPNowPeers(); sendLoRaNbr(1 or 2); broadcastLoRa(); sendSerial(); sendMQTT();
#define ESPNOWG_ACT    sendSerial();
#define LORAG_ACT      sendSerial();
#define SERIAL_ACT     sendESPNowNbr(2); sendESPNowPeers(); sendLoRaNbr(2); broadcastLoRa(); 
#define MQTT_ACT          
#define INTERNAL_ACT   sendSerial();
#define ESPNOW1_ACT    
#define ESPNOW2_ACT    sendSerial();                
#define LORA1_ACT      
#define LORA2_ACT      sendSerial();
```
If I send `[{"id":4,"type":0,"data":1}]` from mqtt explore it shows upp correct in HA but is not in node. Routing issue ? other ?
BTW is there any more information when to use the different kind of routing. I've read back an forth on your side but not 100%  sure how to.
Best Regards Håkan

**Reply by timmbogner** [reply]
Don't worry about it! Sometimes even I get rusty on FDRS specifics, so this is a welcomed refresher! 

The routing you show is for a UART gateway, and if your controllers are connecting properly then you should see data being passed on to them from the serial monitor.  To clarify:

For the data to flow between WiFi (MQTT) and ESP-NOW, you are going to need two gateways connected by serial.

- Gateway0 will be connected to WiFi. It takes the JSON data it receives via MQTT and prints it out on the serial port (UART).

- Gateway1 is connected to that serial port. It takes the JSON data it received via serial (UART) and repackages it and sends it via ESP-NOW or LoRa.

To send test data, open the Arduino serial monitor of Gateway1 and paste in your JSON command. `[{"id":101,"type":1,"data":1}]` should cause your first relay to turn on and four readings to be sent back. Sending it this way is just emulating the messages that Gateway0 would be sending. If you get success with this, then you can switch you efforts to troubleshooting your MQTT/WiFi config on Gateway0.

**Reply by hakha4** [reply]
I **have all above in place and I can see all incoming data both from Lora and espnow nodes in Gateway1 serial monitor and data is sent further via mqtt to Home assistant. SO incoming data works as expected.**

The problem is mqtt out to node. In Serial monitor for Gateway 0 (mqtt) I see that connection is ok both to wifi and mqtt BUT when sending json either from Home assistant or mqtt explorer nothing happens in serial monitor or at node. I don't get it !

When looking in code for fdrs_node there is a line `bool is_controller = false;`. If I understand right a sensor sends data to gateway but not receiving any data  and a controller works both ways,right ?

So,should I set this to true ? or is this variable handled elsewhere . How does system know if a node is a controller or a sensor?

**Reply by timmbogner** [reply]
> So,should I set this to true ? 

No, this is internal stuff that should not be meddled with. At that point in the code, the node is just declaring and initializing the variable `is_controller`, and it will set it to true in `addFDRS()`.

Some questions:
- Is the controller node registering successfully? You will see `Registration accepted. Timeout: 300000` in the Arduino serial monitor of the node.
- Were you able to run the test I described in my last post? That is: if you paste the JSON data to the Arduino serial monitor of Gateway1, does the node respond?
- What specific hardware are you using for the two gateways?

**Reply by hakha4** [reply]
Yes the node is registred ok
No, sending json  from gateway1 don't give any response in node
For gateway1 I have a TTGO BEAM 
![ttgo beam](https://github.com/timmbogner/Farm-Data-Relay-System/assets/3369928/441511a4-3d8f-416b-be60-cad5b0342ae0)

And as mqtt gateway an esp32 board connected with pin 16 rx2 to pin tx on ttgo and pin 17 tx2 to rx on ttgo

**Reply by hakha4** [reply]
I am stupid!! Your questions got me to think about wiring. I have connected to uart0 on ttgo and that must be the issue. I change to uart1 when its daylight! Sorry for taking up your time. I'll get back when I can present a working node with mcp23017. I have a pretty good idea howto now thank's to your patience and help. It's night here in Sweden so I better get some sleep and test when I'm alert!

Best Regards Håkan

**Reply by timmbogner** [reply]
I have it working on my side now. When I enter `[{ id: 102, type: 0, data: 1 }]` into the serial monitor of Gateway1, I get the following. What do you see? Your node is running Irrigation.ino, right? If not, flash that to it as well.
```
17:07:09.360 -> Incoming Serial.
17:07:09.360 -> Sending to ESP-NOW Neighbor #2
17:07:09.360 -> Sending to ESP-NOW peers.
17:07:09.484 -> Incoming ESP-NOW.
17:07:09.484 -> Sending Serial.
17:07:09.484 -> [{"id":101,"type":0,"data":1},{"id":102,"type":0,"data":0},{"id":103,"type":0,"data":1},{"id":104,"type":0,"data":1}]
```

**Reply by timmbogner** [reply]
Gotcha! By default, the data lines are pins 14 and 15 for the original ESP32. You can set the TX and RX to whatever you'd like in the gateway config file, though.

**Reply by hakha4** [reply]
Hi!
Progress but not 100% there yet!

`I have it working on my side now. When I enter [{ id: 102, type: 0, data: 1 }] into the serial monitor of Gateway1, I get the following. What do you see? Your node is running Irrigation.ino, right? If not, flash that to it as well.`

**This works works for me also now after defining new RX/TX pins to the uart GW! BUT I can't send mqtt from Home Assistant or mqtt explorer out TO the irrigation node**

This is my 'test setup'

All nodes have `#define GTWY_MAC      0x01 //Address of the nearest gateway` in code

![fdrs_setup](https://github.com/timmbogner/Farm-Data-Relay-System/assets/3369928/3e883a92-0ae1-49b3-b834-28b42180af20)

1) Data from node 2,3 and 31 are nicely incoming via mqtt to home assistant.

2) I can send data to irrigation node from uart GW and get response as expected 
BUT data are not forwarding to HA and I can't see anything in mqtt GW serial monitor ( am I supposed to do that?)

3) If I send `[{ id: 102, type: 0, data: 1 }] ` from HA or mqtt explorer nothing happens in uart GW monitor or irrigation node

In mqtt GW config look like this

```//  GATEWAY CONFIGURATION
#define FDRS_DEBUG     // Enable USB-Serial debugging
//Addresses
#define UNIT_MAC           0x00  // The address of this gateway

#define ESPNOW_NEIGHBOR_1  0x00  // Address of ESP-NOW neighbor #1
#define ESPNOW_NEIGHBOR_2  0x00  // Address of ESP-NOW neighbor #2
#define LORA_NEIGHBOR_1    0x00  // Address of LoRa neighbor #1
#define LORA_NEIGHBOR_2    0x00  // Address of LoRa neighbor #2

// Interfaces
//#define USE_ESPNOW  
//#define USE_LORA
#define USE_WIFI  // Will cause errors if used with ESP-NOW. Use a serial link instead!
//#define USE_ETHERNET

// Routing
// Options: sendESPNowNbr(1 or 2); sendESPNowPeers(); sendLoRaNbr(1 or 2); broadcastLoRa(); sendSerial(); sendMQTT();
//#define ESPNOWG_ACT    
//#define LORAG_ACT     
#define SERIAL_ACT     sendMQTT();
#define MQTT_ACT       sendSerial();   
#define INTERNAL_ACT   sendMQTT();
#define ESPNOW1_ACT   
#define ESPNOW2_ACT                    
#define LORA1_ACT      
#define LORA2_ACT 

```
I'm still confused about routing, does routing look ok abowe to both send and recieve mqqt data ??

With this setup what would a correct routing be in uart GW and mqtt GW ??

Best Regards Håkan

**Reply by timmbogner** [reply]
Alright, so this may actually be a problem with my documentation... are you sending commands to the gateway under the MQTT topic **"fdrs/command"**? If not, try doing that.

At some point last year I thought this part was broken and topics were ignored, hence issue https://github.com/timmbogner/Farm-Data-Relay-System/issues/167 . Now that I'm looking, it does seem like the gateway filters things based on the MQTT topics listed in fdrs_globals.h as intended. I'll have to figure out why I thought I'd goofed. I certainly recall testing and finding something wrong.

Those routing configs are perfect for an MQTT gateway!

**Reply by hakha4** [reply]
`fdrs/command´` did the trick !!!!. Never thougt of that. Now I can start thinkering for real.

**Reply by timmbogner** [reply]
Awesome! Funny how the solution ended up being in the discussion title. 🤯

I'm sorry the documentation was unclear about that. I'm going to add something that goes into more detail about how MQTT works.

**Reply by hakha4** [reply]
The positive side of struggling with things is that you learn alot at the same time. Once again thank's for all help,appreciated !

**Comment by hakha4** [comment]
I have got a'lot of help from the developer Timm struggling with ESPnow relay nodes and just wanted to do a follow up  and share if other people have had  the same experience and like me are not proffesional programmers.

 Code  below is an example similar to _irrigation node_  example but a little bit easier to change number of relay's needed and with a status LED to easier trobleshoot eventual problems and check when connection got's bad in the field . 

I also have an example using MCP23017  only using 2 pin's (I2C) to get 16 outputs if one needs a'lot of outputs.
 If anyone is interrested in this I can post the code.

 

/*
 FARM DATA RELAY SYSTEM
Developed by Timm Bogner (timmbogner@gmail.com) in Urbana, Illinois, USA.

Inspired from Timm's work and all credit to him!

Sharing this in case someone else beside me can have use for it (_as it is_)


ESP-NOW Sensor Example with various number of relay's just changing a few lines of code
A programmable status  LED (ws2811 single led but  others will work changing type in code) 
to see in field if node is alive and working ok. Can be omitted by just deleting relevant code.

When setting a relay from Front end, the nodes send back and update relay status.
In my case Home assistant but the principle is the same for other Front end's I believe.



Relay's are set a _dirty_ way using _type_ as relay number but works ok, and you only need to define one READING_ID regardless of number of relays.

id = node id(READING_ID), type = relay no (0-3), data = on(0) off(1) inverted mode

Topic: fdrs/command (imortant! (fdrs/data won't work )
  )
  To set relay 0 to on SEND: [{"id":11,"type":0,"data":0}] ON -- INVERTED LOGIC
  To set relay 1 to off SEND: [{"id":11,"type":1,"data":1}] OFF-- INVERTED LOGIC
  "type":x => pin number 0 to 3

compile with DOIT ESP32 DEVKIT V1 Or whatever mcu used but check pin's!

*/

``` cpp
 
#include "fdrs_node_config.h"
#include <fdrs_node.h>
#include <FastLED.h>  // for visual status of node

// programmable status LED
#define NUM_LEDS 1
#define status_PIN 23
CRGB leds[NUM_LEDS];

int tmpdata;             // holds status 0 or 1
int tmp_pinnumber;       // holds relay number
bool isData = false;     // check if data received
bool firsttime = true;   // set all relays to OFF when reboot and update variable's in Frontend
bool connected = false;  // conn status to GW

const int relay_no = 4;  //number of relays to use, add in struct Relay to match pin number and no of relays. 1 to number of avaliable pins

struct Relay {
  int pin;      // Pin number connected to the relay

 //uncomment if relay is set by HIGH or LOW
  bool status;  // Current status of the relay (0: on, 1: off) inverted mode
  //bool status;  // Current status of the relay (0: off, 1: on) non inverted mode
};
//change to number of relays to use and define pin number
Relay relays[relay_no] = {
  { 4, 1 },  // Relay 0 connected to pin 4, initially off if inverted mode
  { 0, 1 },  // Relay 1 connected to pin 0, initially off if inverted mode
  { 2, 1 },  // Relay 2 connected to pin 2, initially off if inverted mode
  { 15, 1 }  // Relay 3 connected to pin 15, initially off if inverted mode
};


//##########################################
void fdrs_recv_cb(DataReading theData) {
  isData = true;
  DBG("Indata");
  tmp_pinnumber = theData.t;
  tmpdata = static_cast<int>(theData.d);  //convert to integer from float
  loadFDRS(tmpdata, tmp_pinnumber);
  //#########################################
}

void setup() {
  // Set relay pins as outputs
  for (int i = 0; i < relay_no; i++) {
    pinMode(relays[i].pin, OUTPUT);
  }
  // initialize status LED
  FastLED.addLeds<WS2811, status_PIN, RGB>(leds, NUM_LEDS);  //change to right type
  FastLED.setBrightness(32);                                 //0 - 255

  beginFDRS();
  if (addFDRS(1000, fdrs_recv_cb)) {
    subscribeFDRS(READING_ID);
    DBG("Connected to Gateway" + GTWY_MAC);
    leds[0] = CRGB::Green;
    FastLED.show();
    connected = true;
  } else {
    DBG("Not Connected to Gateway");
    leds[0] = CRGB::Red;
    FastLED.show();
    connected = false;
  }


  // for debug, prints out a string like: 'off off off off off'   to check in monitor that all works as expecte
  String binaryString = "";
  for (int i = 0; i < relay_no; i++) {
    binaryString += (relays[i].status) ? " Off " : " On ";
  }

  DBG("Relaystatus 0-3 :");
  DBG(binaryString);
  // ##############################################################################
}
void loop() {

  loopFDRS();
  if (!connected) {  //blink status LED to show that we are not connected
    leds[0] = CRGB::Red;
    FastLED.show();
    delay(1000);
    leds[0] = CRGB::White;
    FastLED.show();
    //delay(500);
    pingFDRS(1000);
  }
  if (firsttime && connected) {  //if connected, send all relay's to off to Front end
    firsttime = false;
    for (int i = 0; i < relay_no; i++) {
      loadFDRS(1, i, READING_ID);
      if (sendFDRS()) {
        DBG("OK");
      }

    }  //end loadFDRS loop
  }    //end firsttime

  if (isData) {
    isData = false;

    if (sendFDRS()) {
      leds[0] = CRGB::Blue;  //blink LED to show that command has been sent OK
      FastLED.show();
      delay(300);
      leds[0] = CRGB::Black;
      FastLED.show();
      delay(300);
      leds[0] = CRGB::Green;
      FastLED.show();
      //set relay
      
      // ##########  Set relay to new status #######################
      relays[tmp_pinnumber].status = tmpdata;
      digitalWrite(relays[tmp_pinnumber].pin, relays[tmp_pinnumber].status);
     // ####################################################

      // for debug
      String binaryString = "";
      for (int i = 0; i < relay_no; i++) {
        binaryString += (relays[i].status) ? " Off " : " On ";
      }
      DBG("Relaystatus 0-3 :");
      DBG(binaryString);
    } else {
      //ERROR of some kind, /blink LED to show that command has not been sent
      leds[0] = CRGB::Red;
      FastLED.show();
      delay(300);
      leds[0] = CRGB::Black;
      FastLED.show();
      delay(300);
      leds[0] = CRGB::Yellow;
      FastLED.show();
      DBG("ERROR SENDING DATA");
    }
  }

}  //end void loop
```

Regards Håkan
